pallets/pallet-lottery/src/lib.rs

```diff
@@ -0,0 +1,309 @@
+			ensure_root(origin)?;
+
+			// Ensure the total of Percents adds up to Percent::one()
+			ensure!(Self::check_split_valid(&prize_split), Error::<T>::InvalidPrizeSplitTotal);
+
+			// Set the PrizeSplit storage
+			PrizeSplit::<T>::put(prize_split.clone());
+
+			Self::deposit_event(Event::<T>::PrizeSplitUpdated { split: prize_split });
+			Ok(())
+		}
+
+		/// Set the ticket price.
+		///
+		/// Required Manager or Auditor.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::update_ticket_price())]
+		pub fn update_ticket_price(origin: OriginFor<T>, new_price: T::Balance) -> DispatchResult {
+			let id = ensure_signed(origin)?;
+			let role = T::RoleManager::role(&id);
+			ensure!(
+				role == Some(Role::Manager) || role == Some(Role::Auditor),
+				Error::<T>::IncorrectRole
+			);
+			let old = TicketPrice::<T>::get();
+			// Update the ticket price
+			TicketPrice::<T>::put(new_price);
+			Self::deposit_event(Event::<T>::TicketPriceUpdated { old, new: new_price });
+
+			Ok(())
+		}
+
+		/// Buy numbers of tickets and transfer the funds from the customer account to the lottery
+		/// pool account.
+		///
+		/// Required Customer.
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::buy_ticket())]
+		pub fn buy_ticket(origin: OriginFor<T>, number_of_tickets: u32) -> DispatchResult {
+			// Ensure the caller is a customer account
+			let id = ensure_signed(origin)?;
+			T::RoleManager::ensure_role(&id, Role::Customer)?;
+
+			// Calculate the total price of the tickets
+			ensure!(!TicketPrice::<T>::get().is_zero(), Error::<T>::TicketPriceNotSet);
+			let total_price = TicketPrice::<T>::get().saturating_mul(number_of_tickets.into());
+
+			// Transfer total_price from the customer to the PrizePoolAccount.
+			T::Bank::transfer(&id, &T::PrizePoolAccount::get(), total_price)?;
+
+			// Check if the player exists
+			if let Some(mut tickets) = PlayersAndLotteries::<T>::get(id.clone()) {
+				// Player exists, update the number of tickets
+				tickets = tickets.saturating_add(number_of_tickets);
+				// Update the value in the storage map
+				PlayersAndLotteries::<T>::insert(id.clone(), tickets);
+			} else {
+				// Player doesn't exist, insert a new entry with the given number of tickets
+				PlayersAndLotteries::<T>::insert(id.clone(), number_of_tickets);
+			}
+			Self::deposit_event(Event::<T>::TicketsBought {
+				id: id.clone(),
+				number: number_of_tickets,
+			});
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Checks if the prize split adds up to 100% exactly.
+		fn check_split_valid(split: &[Percent]) -> bool {
+			split.iter().map(|x| *x * 100u32).sum::<u32>() == 100u32
+		}
+
+		/// Increase the random seed.
+		fn next_seed() -> Vec<u8> {
+			let seed = RandomSeed::<T>::get();
+			RandomSeed::<T>::put(seed.wrapping_add(1));
+			seed.encode()
+		}
+
+		/// Payout the won fund to the winner and paid tax to the treasury account.
```

Comment:
/// Resolve the winner of a round of lottery. Transfer `prize - tax` to the winner and tax to the Treasury.